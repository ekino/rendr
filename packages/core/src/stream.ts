import { Readable, Writable } from "stream";
import { Page, createPage } from "./index";
import { Body } from "./types";

/**
 * This function pipe a Readable Stream into a Writable stream in order to pass
 * the response to the server from to the client. This is usefull to act as a proxy
 * and avoid to buffer all the response body in the current process.
 *
 * @param source
 * @param dest
 */
export function pipe(source: Body, dest: Writable): Promise<void> {
  if (typeof source == "string") {
    source = Readable.from(source);
  }

  return new Promise<void>((resolve, reject) => {
    if (!(source instanceof Readable)) {
      return reject("Not a Readable object");
    }

    source.pipe(dest);

    source.on("end", (err: Error) => {
      if (err) {
        reject(err);
      } else {
        resolve();
      }
    });
  });
}

/**
 * This function pipe a Readable Stream into an internal buffer so it is possible
 * to reuse the buffer to create a Page object.
 *
 * @param source
 */
export function pipePageToClient(source: Body): Promise<Page> {
  if (typeof source == "string") {
    source = Readable.from(source);
  }

  let data = "";
  const dest = new Writable({
    write(chunk, _, callback) {
      data += chunk;

      callback();
    },
  });

  return pipe(source, dest).then(() => {
    return createPage(JSON.parse(data));
  });
}
