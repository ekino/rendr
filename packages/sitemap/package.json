{
    "name": "@ekino/rendr-sitemap",
    "license": "MIT",
    "version": "1.0.0-alpha.0",
    "main": "dist/lib/index.js",
    "typing": "dist/lib/index.d.ts",
    "module": "dist/lib-esm/index.js",
    "jsnext:main": "dist/lib-esm/index.js",
    "publishConfig": {
        "access": "public"
    },
    "devDependencies": {
        "@types/node": "^14.6.4"
    },
    "scripts": {
        "build": "tsc && tsc -m es6 --outDir dist/lib-esm && tsc -m amd --outFile dist/_bundles/rendr-core.js"
    },
    "dependencies": {
        "@ekino/rendr-core": "1.0.0-alpha.0",
        "@ekino/rendr-loader": "1.0.0-alpha.0",
        "express": "^4.16.4"
    }
}
