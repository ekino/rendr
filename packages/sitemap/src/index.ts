export * from "./sitemap";
export * from "./types";
