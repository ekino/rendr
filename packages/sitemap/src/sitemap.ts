import {
  PageReference,
  StreamCreator,
  PageReferenceGenerator,
  transformGenerator,
  pipeIteratorToWritable,
} from "@ekino/rendr-core";
import { PageBuilder } from "@ekino/rendr-loader";

import { Writable } from "stream";
import Express from "express";

import { Options } from "./types";
import { ServerResponse } from "http";

export async function sendSitemap(
  res: ServerResponse,
  generator: PageReferenceGenerator
) {
  res.setHeader("Content-Type", "application/xml");
  res.setHeader("X-Rendr-Content-Type", "render/sitemap");

  const iter = transformGenerator(generator(), toSitemapEntry);

  const sitemapWritable = createSitemapWritable((name: string) => res, {
    basePathIndex: "",
  });

  await pipeIteratorToWritable(iter, sitemapWritable);
}

export function createSitemapPageBuilder(
  generator: PageReferenceGenerator
): PageBuilder {
  return async (ctx, page) => {
    // we don't care about the page, we just return binary to the output content
    await sendSitemap(ctx.res, generator);
  };
}

/**
 * This function can be used to expose on sitemap with less than 50K items.
 *
 * If you have a bigger website, creating a sitemap index is suitable, but this cannot
 * be done on the fly.
 *
 * @param generator
 */
export function createSitemapRequestHandler(
  generator: PageReferenceGenerator
): Express.RequestHandler {
  return async (req, res, next) => {
    await sendSitemap(res, generator);
  };
}

/**
 * This function is used as a proxy to open or close file as the
 * sitemap has hard contraints on the size or the number of element
 * in the sitemap.xml file.
 *
 * see https://www.sitemaps.org/protocol.html for mor information.
 *
 * @param createStream
 */
export function createSitemapWritable(
  createStream: StreamCreator,
  options: Options
): Writable {
  const MAX_ENTRY_SITEMAP = 50000;
  const MAX_SIZE_MB = 52428800 - 64;
  const useIndex = options.basePathIndex.length > 0;

  let count = 0;
  let size = 0;
  let index = 1;

  let streamMap = openSitemapStream(createStream, `sitemap_${index}`);
  let streamIndex: Writable;

  if (useIndex) {
    streamIndex = openSitemapIndexStream(createStream);
    streamIndex.write(toSitemapIndexEntry(options, `sitemap_${index}`));
  }

  const w = new Writable({
    write(chunk, encoding, callback) {
      // only apply sitemap specifications if index is on. This feature
      // is usefull for large website (ie, number of page > 50000).
      if (useIndex) {
        count++;
        size += chunk.length;

        if (size > MAX_SIZE_MB || count > MAX_ENTRY_SITEMAP) {
          count = size = 0;
          index++;

          closeSitemapStream(streamMap);
          streamMap = openSitemapStream(createStream, `sitemap_${index}`);
          streamIndex.write(toSitemapIndexEntry(options, `sitemap_${index}`));
        }
      }

      streamMap.write(chunk, encoding);
      callback();
    },
  });

  w.on("finish", () => {
    closeSitemapStream(streamMap);

    if (useIndex) {
      closeSitemapIndexStream(streamIndex);
    }
  });

  return w;
}

/**
 * Tranform an PageReference into a url object used in a sitemap
 *
 * @param ref PageReference
 */
export function toSitemapEntry(ref: PageReference): string {
  if (!ref.settings.lastmod) {
    return "";
  }

  let lastmod = ref.settings.lastmod;

  if (lastmod instanceof Date) {
    lastmod = lastmod.toISOString();
  }

  const settings = {
    lastmod,
    changefreq: ref.settings.changefreq ? ref.settings.changefreq : "weekly",
    priority: ref.settings.priority ? ref.settings.priority : "0.8",
  };

  return `<url><loc>${ref.url}</loc><lastmod>${settings.lastmod}</lastmod><priority>${settings.priority}</priority></url>`;
}

export function toSitemapIndexEntry(options: Options, name: string) {
  return `<sitemap><loc>${
    options.basePathIndex
  }/${name}.xml</loc><lastmod>${new Date().toString()}</lastmod></sitemap>`;
}

function openSitemapStream(
  createStream: StreamCreator,
  name: string
): Writable {
  const stream = createStream(`${name}.xml`);

  stream.write(
    `<?xml version="1.0" encoding="UTF-8"?><urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">`
  );

  return stream;
}

function closeSitemapStream(stream: Writable) {
  stream.write("</urlset>");
  stream.end();
}

function openSitemapIndexStream(createStream: StreamCreator): Writable {
  const stream = createStream("sitemap.xml");

  stream.write(
    `<?xml version="1.0" encoding="UTF-8"?><sitemapindex xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">`
  );

  return stream;
}

function closeSitemapIndexStream(stream: Writable) {
  stream.write("</sitemapindex>");
  stream.end();
}
