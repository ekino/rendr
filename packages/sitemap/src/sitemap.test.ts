import {
  createPageReferencesGenerator,
  createPageReference,
  transformGenerator,
  AsyncPageReferenceGenerator,
  StreamCreator,
  pipeIteratorToWritable,
} from "@ekino/rendr-core";

import { toSitemapEntry, createSitemapWritable } from "./sitemap";

import { Writable } from "stream";

async function* generatorLoop(i: number): AsyncPageReferenceGenerator {
  for (let x = 0; x < i; x++) {
    yield createPageReference(`https://localhost/${x}/${"x".repeat(40)}`, {
      lastmod: "2005-01-01",
    });
  }
}

describe("Test sitemaps code", () => {
  test("test toSitemapEntry with date", () => {
    const ref = createPageReference("https://myurl.com/", {
      lastmod: new Date("2019-10-24T14:16:29.329Z"),
    });

    const s = toSitemapEntry(ref);

    expect(s).toMatchSnapshot();
  });

  test("test sitemap with index file", async () => {
    const generator = createPageReferencesGenerator({
      group_1: () => generatorLoop(40000),
      group_2: () => generatorLoop(50000),
    });

    let call = 0;

    const files: string[] = [];
    const iter = transformGenerator(generator(), toSitemapEntry);

    const streamCreator: StreamCreator = (name: string) => {
      files.push(name);

      const stream = new Writable({
        write(chunk, encoding, callback) {
          call++;
          callback();
        },
      });

      return stream;
    };

    const sitemapWritable = createSitemapWritable(streamCreator, {
      basePathIndex: "https://localhost",
    });

    await pipeIteratorToWritable(iter, sitemapWritable);

    expect(files).toMatchSnapshot();
    // 4 is coming from open and close of the sitemap file
    // 4 is coming from open and close of the sitemap index file
    expect(call).toBe(90000 + 4 + 4);
  });

  test("test sitemap with no index file", async () => {
    const generator = createPageReferencesGenerator({
      group_1: () => generatorLoop(10),
      group_2: () => generatorLoop(5),
    });

    let buffer = "";

    const files: string[] = [];
    const iter = transformGenerator(generator(), toSitemapEntry);

    const streamCreator: StreamCreator = (name: string) => {
      files.push(name);

      const stream = new Writable({
        write(chunk, encoding, callback) {
          buffer += chunk;
          callback();
        },
      });

      return stream;
    };

    const sitemapWritable = createSitemapWritable(streamCreator, {
      basePathIndex: "",
    });

    await pipeIteratorToWritable(iter, sitemapWritable);

    expect(files).toMatchSnapshot();
    expect(buffer).toMatchSnapshot();
  });
});
