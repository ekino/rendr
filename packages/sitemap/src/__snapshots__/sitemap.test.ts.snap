// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test sitemaps code test sitemap with index file 1`] = `
Array [
  "sitemap_1.xml",
  "sitemap.xml",
  "sitemap_2.xml",
]
`;

exports[`Test sitemaps code test sitemap with no index file 1`] = `
Array [
  "sitemap_1.xml",
]
`;

exports[`Test sitemaps code test sitemap with no index file 2`] = `"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\"><url><loc>https://localhost/0/xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx</loc><lastmod>2005-01-01</lastmod><priority>0.8</priority></url><url><loc>https://localhost/1/xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx</loc><lastmod>2005-01-01</lastmod><priority>0.8</priority></url><url><loc>https://localhost/2/xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx</loc><lastmod>2005-01-01</lastmod><priority>0.8</priority></url><url><loc>https://localhost/3/xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx</loc><lastmod>2005-01-01</lastmod><priority>0.8</priority></url><url><loc>https://localhost/4/xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx</loc><lastmod>2005-01-01</lastmod><priority>0.8</priority></url><url><loc>https://localhost/5/xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx</loc><lastmod>2005-01-01</lastmod><priority>0.8</priority></url><url><loc>https://localhost/6/xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx</loc><lastmod>2005-01-01</lastmod><priority>0.8</priority></url><url><loc>https://localhost/7/xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx</loc><lastmod>2005-01-01</lastmod><priority>0.8</priority></url><url><loc>https://localhost/8/xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx</loc><lastmod>2005-01-01</lastmod><priority>0.8</priority></url><url><loc>https://localhost/9/xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx</loc><lastmod>2005-01-01</lastmod><priority>0.8</priority></url><url><loc>https://localhost/0/xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx</loc><lastmod>2005-01-01</lastmod><priority>0.8</priority></url><url><loc>https://localhost/1/xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx</loc><lastmod>2005-01-01</lastmod><priority>0.8</priority></url><url><loc>https://localhost/2/xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx</loc><lastmod>2005-01-01</lastmod><priority>0.8</priority></url><url><loc>https://localhost/3/xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx</loc><lastmod>2005-01-01</lastmod><priority>0.8</priority></url><url><loc>https://localhost/4/xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx</loc><lastmod>2005-01-01</lastmod><priority>0.8</priority></url></urlset>"`;

exports[`Test sitemaps code test toSitemapEntry with date 1`] = `"<url><loc>https://myurl.com/</loc><lastmod>2019-10-24T14:16:29.329Z</lastmod><priority>0.8</priority></url>"`;
